src/java.base/share/classes/java/io/Console.java

```diff
@@ -592,25 +595,43 @@ public int read(char[] cbuf, int offset, int length)
 
         CHARSET = cs;
 
+        cons = instantiateConsole(istty);
+
         // Set up JavaIOAccess in SharedSecrets
         SharedSecrets.setJavaIOAccess(new JavaIOAccess() {
             public Console console() {
-                if (istty) {
-                    if (cons == null)
-                        cons = new Console();
-                    return cons;
-                }
-                return null;
-            }
-
-            public Charset charset() {
-                return CHARSET;
+                return cons;
             }
         });
     }
-    private static Console cons;
+
+    @SuppressWarnings("removal")
+    private static Console instantiateConsole(boolean istty) {
+        try {
+            // Try loading providers
+            PrivilegedAction<Console> pa = () -> {
+                var consModName = System.getProperty("jdk.console",
+                        JdkConsoleProvider.DEFAULT_PROVIDER_MODULE_NAME);
+                return ServiceLoader.load(JdkConsoleProvider.class).stream()
+                        .map(ServiceLoader.Provider::get)
```

Comment:
Furthermore, I think in its current form it means that this will load/instantiate any `JdkConsoleProvider` implementations that are accessible to the thread context classloader but may not have been from the module configured through `jdk.console` system property. That could potentially mean, in the best case, unnecessary classloading of additional classes and in the worst case, could result in `ServiceLoader.Provider::get` throwing a `ServiceConfigurationError` error for any of such unused provider implementations, thus forcing us to use `java.io.Console` instance.

Reply:
> Furthermore, I think in its current form it means that this will load/instantiate any `JdkConsoleProvider` implementations that are accessible to the thread context classloader but may not have been from the module configured through `jdk.console` system property. That could potentially mean, in the best case, unnecessary classloading of additional classes and in the worst case, could result in `ServiceLoader.Provider::get` throwing a `ServiceConfigurationError` error for any of such unused provider implementations, thus forcing us to use `java.io.Console` instance.

You are right that the ServiceLoader.load should specify the system class loader or the boot layer. However, there isn't an accessibility issue as a class loader just load classes so it's more about visibility and whether the TCCL will ultimately delegate to the application class loader.

Reply:
`module-info.java` in the java.base only allows `jdk.internal.le` and `jdk.jshell` modules to access the `jdk.internal.io.JdkConsoleProvider` interface. So unless the user intentionally exports it, no other implementations are effectively instantiated.